File: projects/m1/035-the-twelve-days-of-christmas/solutions/enricozoboli/solution_the_twelve_days_of_christmas.py
def make_ordinal(num_day):
    suffix = "th"
    if num_day == 1:
        ordinal_num = "first"
    elif num_day == 2:
        ordinal_num = "second"
    elif num_day == 3:
        ordinal_num = "third"
    elif num_day == 4:
        ordinal_num= "four" 
    elif num_day == 5:
        ordinal_num = "fif"
    elif num_day == 6:
        ordinal_num = "six"
    elif num_day == 7:
        ordinal_num = "seven" 
    elif num_day == 8:
        ordinal_num = "eigh"
    elif num_day == 9:
        ordinal_num = "nin"
    elif num_day == 10:
        ordinal_num = "ten"
    elif num_day == 11:
        ordinal_num= "eleven"
    else:
        ordinal_num = "twelf"
    
    if 1<= num_day <= 3:
        return ordinal_num
    else: 
        return ordinal_num + suffix

def func_verse(num_day):
    if num_day == 1:
        verse = "A partridge in a pear tree."
        return verse
    else:
        if num_day == 2:
            verse = "Two turtle doves,"
        elif num_day == 3:
            verse = "Three French hens,"
        elif num_day == 4:
            verse= "Four calling birds," 
        elif num_day == 5:
            verse = "Five gold rings,"
        elif num_day == 6:
            verse = "Six geese a-laying,"
        elif num_day == 7:
            verse = "Seven swans a-swimming," 
        elif num_day == 8:
            verse = "Eight maids a-milking,"
        elif num_day == 9:
            verse = "Nine ladies dancing,"
        elif num_day == 10:
            verse = "Ten lords a-leaping,"
        elif num_day == 11:
            verse= "Eleven pipers piping,"
        elif num_day == 12:
            verse = "Twelve drummers drumming,"
    return f"{verse}\n"




def main():
    total_verse = ""
    last_verse = "And a partridge in a pear tree."
    for i in range(1,13):
        if i == 1:
            print(f"On the {make_ordinal(i)} day of Christmas"
            f"my true love sent to me:")
            print(func_verse(i))
        else:
            print(f"\nOn the {make_ordinal(i)} day of Christmas"
            f"my true love sent to me:")
            if not total_verse:
                print(func_verse(i) + last_verse)
            else:
                print(total_verse + func_verse(i) + last_verse)
            total_verse += func_verse(i)
        


if __name__ == "__main__":
    main()

File: projects/m1/035-the-twelve-days-of-christmas/solutions/valentinaBazzacchi/solution-the-twelve-days-of-christmas.js
function getSongVerse(number) { // verseNumber[number-1] and items[number-1] allow to select the different elements through the for loop 
    let verseNumber = ["first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth"];
    let versePart1 = "On the " + verseNumber[number-1] + " day of Christmas my true love sent to me: \n"
    items = ["And a partridge in a pear tree.", "Two turtle doves,", "Three french hens,", "Four calling birds,", "Five golden rings,", "Six geese a-laying,", 
    "Seven swans a-swimming,", "Eight maids a-milking,", "Nine ladies dancing,", "Ten lords a-leaping,", "Eleven pipers piping,", "Twelve drummers drumming,"];
    items = items[number-1];
    // Condition for handling the exception of verse number 1
    if (number == 1) { 
        return versePart1 + "A partridge in a pear tree.\n";
    } else {
        return versePart1 + items + "\n";
    }
}     
// The following variables are declared in the global scope because they will be used in code outside the function
let items;
let prevItems = [];
let song = "";
// The iterator "i" become the argument of the function getSongVerse(number).
for (let i = 1; i < 13; i++) {
    prevItems.unshift(items);
    song += (getSongVerse(i) + prevItems.join("\n")) + "\n";
}

console.log(song);

 /* 
 When i = 1 and before the function call (code line n. 20), items = undefined so the array prevItems remains empty. 
 --> After the function call (code line 21), items become [0] but the program follow the if condition (number == 1) within the function 
 and return VersePart1 (with verseNumber[0]) + the string defined as an exception "A partridge in a pear tree.\n". 
 The array prevItems is still empty and is not displayed on the screen! 
 
 When i = 2 the current value of items is [0] and it is added to the beginning of the array prevItems (at the line code n. 20),
 --> After the function call (code line 21), it returns VersePart1 (with verseNumber[1]) + items become [1] + prevItem which now contains item[0].
 
 When i = 3 the current value of items is [1] and it is added to the beginning of the array prevItems[items [1], items[0]] (at the line code n. 20),
--> After the function call (code line 21), it returns VersePart1 (with verseNumber[2]) + items become [2] + prevItem which now contains items [1], item[0].
 
And so on until i = 12 ...
 */